### PR TITLE
refactor: split receipts and events, use `bincode` for storage

### LINKED ISSUE #1843: Split receipts and events
Currently, `starknet_getEvents` deserializes the entire receipt just to get the events. Now that we're adding P2P sync, we are storing receipts without events during sync. This should not cause any problems if we are careful during sync, but isn't exactly the most elegant approach.

Additionally, it would be useful to start storing transactions, receipts, and events via bincode to save storage. (Currently they are stored as JSON). This will require an expensive database migration, but at this point it seems we can't avoid it.

### PR DESCRIPTION
**Pull request by sistemd**
This mega PR is the result of many smaller PRs over the past week. It separates receipts and events and their storage, and stores them via bincode instead of JSON.

Closes #1843.

Since all previous PRs have already been approved, for this one it should be sufficient to just skim the changes and more or less "rubber-stamp" the approval.

### CODE REVIEW

File: Cargo.toml
Comment (Mirko-von-Leipzig) [inline Cargo.toml:117]: Think this is gone now?

Comment (CHr15F0x) [inline Cargo.toml:117]: iirc `smallvec` is removed on `main`: https://github.com/eqlabs/pathfinder/pull/1895

**Review by Mirko-von-Leipzig — APPROVED** [review]
Think lets wait for everyone's approval just to be safe.

Do we have some final numbers for storage savings?

**Review by CHr15F0x — APPROVED** [review]
LGTM

File: crates/storage/src/schema/revision_0052.rs
Comment (kkovaacs) [inline crates/storage/src/schema/revision_0052.rs:171]: Hmm, does this preserve the `starknet_transactions_block_hash` index we had for the old table? I think it doesn't.

File: crates/storage/src/schema/revision_0052.rs
Comment (kkovaacs) [inline crates/storage/src/schema/revision_0052.rs:12]:
```suggestion
    tracing::info!("Migrating starknet_transactions to new format");
```

File: crates/storage/src/schema/revision_0052.rs
Comment (kkovaacs) [inline crates/storage/src/schema/revision_0052.rs:143]:
```suggestion
                    "Migrating transactions: {:.2}%",
```

File: crates/storage/src/schema/revision_0052.rs
Comment (kkovaacs) [reply crates/storage/src/schema/revision_0052.rs:171]: Yep, it's missing after the migration.

The whole migration took 28 minutes on my mainnet db. Re-creating the missing index (`CREATE INDEX starknet_transactions_block_hash ON starknet_transactions(block_hash)`) is then another 2.5 minutes (I've done that manually).

Comment (Mirko-von-Leipzig) [reply crates/storage/src/schema/revision_0052.rs:171]: Actually this reminds me. Do we want to change this block hash to a block number index instead? It would also be faster to lookup..

**Review by Mirko-von-Leipzig — APPROVED** [review]
LGTM! Nice work 😀

**Pull request merged**
